README.md
# USE AT OWN RISK, NOT ALL SCRIPTS ARE VERIFIED TO WORK 100% WITH LTO
# 
# LTO_LPoSDistributor
A revenue distribution tool for LTO.network nodes

Welcome to LTO.network version of the LPoSdistribution script,
which queues up multiple sessions and automates next session info :-)

Massive thank you to Rob / G1zm0 (http://dev.pywaves.org/LTO/generators/) and Sven Stam from LTO team for helping out and fixing bugs.

Many thanks to the version of Plukkie (https://github.com/plukkie/WavesLPoSDistributer), and the original version of Marc Jansen (https://github.com/jansenmarc/WavesLPoSDistributer) and the fork of W0utje (https://github.com/w0utje/WavesLPoSDistributer)!

You can lease to '3JqGGBMvkMtQQqNhGVD6knEzhncb55Y7JJ5' for Liquid Leasing Network or to Rob/G1zm0: '3JeUGgoCUy5wXpNKHqhaLpvGZrshtvwt9b9'

## Installation steps: prerequisits
First of all, you need to install Node.js (https://nodejs.org/en/) and npm.\
This version is succesfully tested with versions;
 - node v10.12.0 (allthough lower should work probably)
 - npm 6.4.1 (allthough lower should work probably)
 - tested on Ubuntu 14.0 with kernel 4.4.0-116-generic (allthough of minor importance)
 - get the latest version from github: git clone https://github.com/jayjaynl/LTO_LPoSDistributor.git

To install node.js and npm, do following steps;
 - add repository: curl -sL https://deb.nodesource.com/setup_10.x | sudo bash -
 - install both packages: sudo apt-get install -y nodejs

Now you can proceed with the lto LPOS scripts installation, select the variant which applies to you.

## Installation steps: first time users
These steps are for users that do not use an older version of the LPoSdistributer package yet.
1. CD into the LPoS package directory : LTO_LPoSDistributer
2. install the package independencies:
```sh
mkdir node_modules
npm install
```
3. configure for one time only the initial settings of the relevant blocks in the batchinfo.json file:
```sh
EDIT file batchinfo.json with vim or nano;

{
    "batchdata": {
        "paymentid": "1",				<== Leave as is
        "paystartblock": "1044012",			<== Put here same value as 'scanstartblock'. It's when payouts should start
        "paystopblock": "1050000",			<== Put here a value when payouts should stop (i.e. paystartblock+5000)
							    It doesn't really matter, as long as it is higher than paystartblock.
							    It only counts for the first run, and if no blocks were forged yet, that
							    is no problem. Follow up session results are just queued up in line :-))
        "scanstartblock": "1044012"			<== Put here the blockheight of the first ACTIVE lease
    }
}
```
NOTE\
This file is updated automatically after the collector session finishes.\
The size of the next batch (paystart / paystop blocks), is used from the\
'blockwindowsize' configuration value in the appng.js file.

4. EDIT file appng.js with vim or nano;
   This file is the collector that checks all blocks for leases and fees;
```sh
const myleasewallet = '<your node wallet>';		<== Put here the address of the wallet that your node uses
const myquerynode = "http://localhost:6869";		<== The node and API port that you use (defaults to localhost)
const feedistributionpercentage = 90;			<== How many % do you want to share with your leasers (defaults to 90%)
const blockwindowsize = 10000;				<== How many blocks to process for every subsequent paymentcycle.

var nofeearray = [ ];					<== Put here wallet addresses that you want to exclude from payments,
							    Default empty, so everyone get's payouts
```
5. EDIT file checkPaymentsFile.js with vim or nano;
```sh
var config = {
    <SNIP>,
    node: 'http://localhost:6869',			<== Change this value to your blockchain node/API port (defaults to localhost)
    <SNIP>
};
```
6. EDIT file **massPayment.js** and **masstx.js**
```sh
var config = {
    <SNIP>,
    node: 'http://localhost:6869',			<== Change this value to your blockchain node/API port (defaults to localhost)
    apiKey: 'your api key'				<== Put here the API key of your lto node
};
```
NOTE\
For security reasons, remove 'rwx' worldrights from massPayment.js and masstx.js -> ```chmod o-rwx massPayment.js``` \
Repeat for masstx.js. Now you can jump to chapter "Running the collector sessions".

## Running the collector sessions
After a successful configuration of the tool, start with:
```sh
node appng.js OR start_collector.sh
```
In Windows it's easier to just run:
```sh
node --stack-size=65565 --max-old-space-size=8192 appng.js
```
NOTE0\
If you can't start 'start_collector', check if the script has execute 'x' on it.\
If not add with: ```chmod u+x start_collector.sh```

NOTE1\
The script can consume a serious amount of memory and exits with errors during it's run.\
Therefore I've put 'start_collector.sh' script as starter which runs 'node appng.js' with some memory optimized settings.\
For me it works with tweaks to 65KB of stack memory and 8GB of available RAM. So use 'start_collector.sh' if you run into problems\
and tweak to your available RAM. If it keeps on exitting, then shrink the block batchsize that are collected in one batch.\
This way multiple smaller batchsizes will be collected and consume less memory.\
To decrease the initial batchsize, edit file 'batchinfo.json' and set 'paystopblock' smaller (closer to 'paystartblock').
To have all subsequent runs also changed, edit file 'appng.js' and set 'blockwindowsize' smaller.

NOTE2\
To run the collector tool every night @1 AM, edit /etc/crontab and put in following line;\
```sh
00 01 * * * root cd /home/myuser/ltoLPoSDistributer/ && ./start_collector.sh
```
After the tool ran, it finishes up by writing the actual payments to be done into the file which is configured in the script by:
```sh
var config = {
	filename: 'ltoleaserpayouts',
```
The name is constructed together with the paymentid (or batchID) of every batch session.\
So, for the first run, the following three files will be created;
- ltoleaserpayouts1.json
- ltoleaserpayouts1.html
- ltoleaserpayouts1.log

The batchID is added to the payqueue.dat file. When there are already pending payments, it's just added.
For the next session, the batchid is incremented by 1 and the batchdata.json file is updated with the new blockheights and batchID.

## Checking pending payments
After the collector ran (or ran multiple times as you wish), you can check the payments that are stored in the payment queue.
The script for checking is checkPaymentsFile.js. After you configured some settings (see above), you can start with;
```sh
node checkPaymentsFile.js
```
The script reads all all batchIDs from the payqueue.dat file and the corresponding leaser files that were constructed by the collector tool. It does only checking, nothing else. The results for all pending payments is printed on the screen.\
The checker also calculates the cost for single transactions (payment tool massPayment.js) and the cost for masstransfers (masstx.js).
Often the number of transactions are high enough to benefit from masstransfers, which are often cheapest :-)
See more about both payment psossibilities in next chapter (doing the payments).
After checking this information, you have a good overview what tokens and the amounts are planned for payout and which transaction type is best to use!

## Doing the payments
For the actual payout, you can choose the massPayment.js tool or the masstx.js tool. They can be started with:
```sh
node massPayment.js or node masstx.js
```
The massPayment.js tool does a single transaction for every payment to be done.
The masstx.js tool makes use of masstransfers and pushes multiple payments for one one and the same asset
into one masstransfer transaction. This optimizes blockchain storage and and transaction costs.
If you run the checker first (checkPaymentsfile.js), you'll get a nice overview which method is cheapest for
for your payment batches. Both tools can just be used interchangelly.
All batchIDs are sequencially read from the payment queue and the transactions are executed.\
When a job finishes, the batchID is removed from the payqueue.dat and the three ltoleaserpayoutX.* files\
are moved to the archival directory (paymentsDone/).

NOTE massPayment.js\
If there would be a crash of the system, script or other transaction breaking interruption,\
make note of the last succesfull transaction counter and the batchID. Then edit the massPayment.js\
file and change these values for:
```sh
const crashconfig = {
        batchidstart: '0',		<== batchID here
        transactionstart: '0' }		<== last succesfull transaction +1
```
Then start the 'node massPayment.js'.\
The values you can leave in or you can put it back to 0 / 0 if you like.

## Why three seperate tools?
We decided to use seperate tools since this allows for additional tests of the payments before the payments are actually executed.
On the other hand, it does not provide any drawback since both scripts could also be called directly one after the other with:
```sh
node appng.js && node masstx.js or ./start_collector && node massPayment.js
```
However, it is strongly recommended to check the payments before the actual payments are done.\
So what you could do for example, run from crontab;
- run the collector session every saterday evening
- run the checkPaymentFile every sunday, mail the output, so you have visibility
- run the massPayment job every tuesday

With this scheme, you have a nice automated schema and works as follows;
- Every wednesday & saterday, if the collector batchsize is still too large (because the mainnet blockheight is to low),
  the cronjob exits and waits till next collector day. If the blockrange fits, the payments are collected and
  the job is logged and queued
- Every Sunday, the checkPayment job checks the payqueue. If empty, it exits. If there is (are) job(s) in the queue,
  the paymentdata for all batchIDs are shown and your output was send by email. You have time to check the results
- Every tuesday, the payment is done for all batchIDs in the queue.

NOTE\
It safe to schedule the collector and check jobs. However, regarding payments,\
it's always possible that something disrupts the transaction process, in which payments\
could fail and leasers don't receive what they should. It's up to you, if you feel confident\
with automated payments. If not, you can just execute massPayment.js by hand.\
MassPayment has forseen in the event that crashes or failed transactions (due to whatever reason) happen,\
by which you can add the batchID and the number of the last succesfull transactionnr.+1, to the file\
and then start massPayment.js again. Transactions will be executed from where the failures started.\
The nice thing is that the three tools are decoupled. So, if you run the collector three times a week\
and the the checks every week and the payout just once a month or whenever you feel it's a good moment,\
that's all fine. It also depends on the frequency of blockhits for your node and the blockwindows size\
you configure. It's all up to you and it doesn't bite one another.

## Disclaimer
Please always test your resulting payment scripts, e.g., with the _checkPaymentsFile.js_ script!


CHANGELOG.txt
/* =====lto====lto====lto====lto====lto====lto====lto====lto====lto====lto===lto====lto====
 *
 * This is a fork by of Plukkie's version of the LPoSDistributor
 *
 * Massive thank you to Rob / G1zm0 (http://dev.pywaves.org/LTO/generators/) and Sven Stam from LTO team for helping out and fixing bugs.
 *
 * Many thanks to the version of Plukkie, and the original version of Marc Jansen and the fork of W0utje.
 *
 * You can lease to '3JqGGBMvkMtQQqNhGVD6knEzhncb55Y7JJ5' for Liquid Leasing Network
 * 
 * or to Rob/G1zm0: '3JeUGgoCUy5wXpNKHqhaLpvGZrshtvwt9b9'
 *
 * Thank you and Enjoy!
 *
/* =====lto====lto====lto====lto====lto====lto====lto====lto====lto====lto===lto====lto====

================================================================================================================
 * V1.0.0
 * Liquid Leasing Network - LTO_LPoSDistributor
 *
 * CHANGES
 * - Removed all 'Waves' and 'MRT' references
 * - Rewrote some instructions
 * - Rewrote changelog
 * - Rewrote readme
 * - Rewrote appng, checkpaymentfile and masstx scripts to work with LTO
 *
 * ================================================================================================================= */



appng.js
/* =================================================================================================================
 * V1.0.0 - LTO_LPoSDistributor
 *
 * A fork of Plukkies version of the LPoSdistribution script for LTO.network (https://github.com/plukkie/WavesLPoSDistributer)
 * 
 * Massive thank you to Rob / G1zm0 (http://dev.pywaves.org/LTO/generators/) and Sven Stam from LTO team for helping out and fixing bugs. 
 * 
 * You can lease to '3JqGGBMvkMtQQqNhGVD6knEzhncb55Y7JJ5' for Liquid Leasing Network
 * 
 * or to Rob/G1zm0: '3JeUGgoCUy5wXpNKHqhaLpvGZrshtvwt9b9'
 *
 * Please see CHANGELOG.txt for all features/changes
 * Please see README.md for complete explanation of LPoSdistributor package
 *
 * Bare minimum changes to edit:
 * - put your address here: const myleasewallet = '<Put your Leasewallet address here>';
 * - put the blockchain node and port where you access the API: const myquerynode = "http://<node ip/name:port here>";
 * - How many percentage fees do you want to share with your leasers: const feedistributionpercentage = <nr here>;
 * - How many blocks do you want to collect in every collector run: const blockwindowsize = <nr here>;
 *
 * Don't forget to set values in batchinfo.json only once! :-)
 * See README.md for explanation
 *
 * ================================================================================================================= */


// START - Put your settings here
const myleasewallet = 'your address';	//Put here the address of the wallet that your node uses
const myquerynode = "http://localhost:6869";	//The node and API port that you use (defaults to localhost)
const feedistributionpercentage = 95;		//How many % do you want to share with your leasers (defaults to 90%)
const blockwindowsize = 5000; 			//how many blockss to proces for every paymentcycle

// Put here wallet addresses that will receive no fees
// var nofeearray = [ "3P6CwqcnK1wyW5TLzD15n79KbAsqAjQWXYZ",       //index0
//                    "3P9AodxBATXqFC3jtUydXv9YJ8ExAD2WXYZ" ];
var nofeearray = [ ]; 
// END - your settings



var request = require('sync-request');
var fs = require('fs');

// file with the batch data to start collecting. Will be updated after succesfull appng run
var batchinfofile = "batchinfo.json";
var payqueuefile    = "payqueue.dat";

if (fs.existsSync(batchinfofile)) {

   var rawbatchinfo = fs.readFileSync(batchinfofile);
   var batchinfo = JSON.parse(rawbatchinfo);
  
   mybatchdata = batchinfo["batchdata"];
   paymentstartblock = parseInt(mybatchdata["paystartblock"]);
   paymentstopblock = parseInt(mybatchdata["paystopblock"]);
   startscanblock = parseInt(mybatchdata["scanstartblock"]);
   payid = parseInt(mybatchdata["paymentid"]); 

   // Collect height of last block in lto blockchain
   let options = {
	uri: "/blocks/height",
	baseUrl: myquerynode,
	method: "GET",
	headers: {
	json: true
	}
   };
   
   let blockchainresponse = request(options.method, options.baseUrl + options.uri, options.headers)
   let lastblockheight = parseInt(JSON.parse(blockchainresponse.body).height) 
   
   if (paymentstopblock > lastblockheight) {
	let blocksleft = paymentstopblock - lastblockheight
        console.log("\n Current blockheight is " + lastblockheight + ". Waiting to reach " + paymentstopblock + " for next payment round.")
        console.log(" This is approximaly in ~" + Math.round((blocksleft)/60) + " hrs (" + (Math.round((blocksleft/60/24)*100))/100 + " days).\n")
        return;
   } else { var backupbatchinfo = fs.writeFileSync(batchinfofile + ".bak",fs.readFileSync(batchinfofile)) }  //Create backup of batchdatafile

} 
else {
     console.log();
     console.log("Error, batchfile",batchinfofile,"missing. Will stop now.");
     console.log();
     return; //if the batchinfofile doesn't exist stop further processing
}

var config = {
    address: myleasewallet,
    startBlockHeight: paymentstartblock,
    endBlock: paymentstopblock,
    filename: 'ltoleaserpayouts', //.json added automatically
    paymentid: payid,
    node: myquerynode,
    feeAmount: 25000000,
    percentageOfFeesToDistribute: feedistributionpercentage
};

var myLeases = {};
var myCanceledLeases = {};

var currentStartBlock = startscanblock;

var fs=require('fs');
var prevleaseinfofile = config.startBlockHeight + "_" + config.address + ".json";
if (fs.existsSync(prevleaseinfofile))
{
	console.log("reading " + prevleaseinfofile + " file");
	var data=fs.readFileSync(prevleaseinfofile);
	var prevleaseinfo=JSON.parse(data);
	myLeases = prevleaseinfo["leases"];
	myCanceledLeases = prevleaseinfo["canceledleases"];
	currentStartBlock = config.startBlockHeight;
}

//do some cleaning
var cleancount = 0;
for(var cancelindex in myCanceledLeases)
{
    if(myCanceledLeases[cancelindex].leaseId in myLeases)
    {
        //remove from both arrays, we don't need them anymore
        delete myLeases[cancelindex];
        delete myCanceledLeases[cancelindex];
        cleancount++;
    }

}
console.log("done cleaning, removed: " + cleancount);

var payments = [];

var myAliases = [];

var BlockCount = 0;

var LastBlock = {};

var myForgedBlocks = [];

/**
  * This method starts the overall process by first downloading the blocks,
  * preparing the necessary datastructures and finally preparing the payments
  * and serializing them into a file that could be used as input for the
  * masspayment tool.
 */
var start = function() {
//   console.log('get aliases');
//   myAliases = getAllAlias();
    console.log('getting blocks...');
    var blocks = getAllBlocks();
    console.log('preparing datastructures...');
    prepareDataStructure(blocks);
    console.log('preparing payments...');
    myForgedBlocks.forEach(function(block) {
        if (block.height >= config.startBlockHeight && block.height <= config.endBlock) {
            var blockLeaseData = getActiveLeasesAtBlock(block);
            var activeLeasesForBlock = blockLeaseData.activeLeases;
            var amountTotalLeased = blockLeaseData.totalLeased;

            distribute(activeLeasesForBlock, amountTotalLeased, block);
            BlockCount++;
        }
    });
    //Get last block
    LastBlock = blocks.slice(-1)[0] ;

    pay();
    console.log("blocks forged: " + BlockCount);
};

/**
 * This method organizes the datastructures that are later on necessary
 * for the block-exact analysis of the leases.
 *
 *   @param blocks all blocks that should be considered
 */

var prepareDataStructure = function(blocks) {

    blocks.forEach(function(block,index) {
    var checkprevblock = false;
	var myblock = false;
        var ltoFees = 0;

        if (block.generator === config.address)
        {
            myForgedBlocks.push(block);
            checkprevblock = true;
            myblock = true;
        }
	var blockltofees=0;

        block.transactions.forEach(function(transaction)
        {
            // type 8 are leasing tx
            if (transaction.type === 8 && transaction.recipient === config.address){
                transaction.block = block.height;
                myLeases[transaction.id] = transaction;
            } else if (transaction.type === 9 && myLeases[transaction.leaseId]) { // checking for lease cancel tx
                transaction.block = block.height;
                myCanceledLeases[transaction.leaseId] = transaction;
            }

            if(myblock)
            {
                // considering lto fees
                if(transaction.fee < 150000000000) // if tx lto fee is more dan 150 lto, filter it. probably a mistake by someone
                {
                    //ltoFees += (transaction.fee*0.4);
                    blockltofees += transaction.fee;
                } else {
                    console.log("Filter TX at block: " + block.height + " Amount: " +  transaction.fee)
                }
            }
      });
      ltoFees += Math.round(parseInt(blockltofees / 5) * 2);

      blockltofees=0;

      if(checkprevblock)
      {
        if (index > 0)
        {
            //console.log("Next: " + blocks[index + 1]);
            var prevblock = blocks[index - 1];
            prevblock.transactions.forEach(function(transaction)
            {
                // considering lto fees
                if(transaction.fee < 150000000000) // if tx lto fee is more dan 150 lto, filter it. probably a mistake by someone
              	{
                  	//ltoFees += (transaction.fee*0.6);
                  	blockltofees += transaction.fee;
                } else {
  			        console.log("Filter TX at block: " + block.height + " Amount: " +  transaction.fee)
  		        }
            });
        }

      ltoFees += (blockltofees - Math.round(parseInt(blockltofees / 5) * 2));

      }

        block.ltoFees = ltoFees;

    });
};

/**
 * Method that returns all relevant blocks.
 *
 * @returns {Array} all relevant blocks
 */
var getAllBlocks = function() {

    var blocks = [];

    while (currentStartBlock < config.endBlock) {
        var currentBlocks;

        if (currentStartBlock + 99 < config.endBlock) {
            console.log('getting blocks from ' + currentStartBlock + ' to ' + (currentStartBlock + 99));
            currentBlocks = JSON.parse(request('GET', config.node + '/blocks/seq/' + currentStartBlock + '/' + (currentStartBlock + 99), {
                'headers': {
                    'Connection': 'keep-alive'
                }
            }).getBody('utf8'));
        } else {
            console.log('getting blocks from ' + currentStartBlock + ' to ' + config.endBlock);
            currentBlocks = JSON.parse(request('GET', config.node + '/blocks/seq/' + currentStartBlock + '/' + config.endBlock, {
                'headers': {
                    'Connection': 'keep-alive'
                }
            }).getBody('utf8'));
        }
        currentBlocks.forEach(function(block)
        {
            if (block.height <= config.endBlock) {
                blocks.push(block);
            }
        });

        if (currentStartBlock + 100 < config.endBlock) {
            currentStartBlock += 100;
        } else {
            currentStartBlock = config.endBlock;
        }
    }

    return blocks;
};

/**
 * This method distributes lto fees
 * the given block.
 *
 * @param activeLeases active leases for the block in question
 * @param amountTotalLeased total amount of leased lto in this particular block
 * @param block the block to consider
 */
var distribute = function(activeLeases, amountTotalLeased, block) {

    var fee = block.ltoFees;

    for (var address in activeLeases) {

	if ( nofeearray.indexOf(address) == -1 ) {	// leaseaddress is not marked as 'no pay address'
		var share = (activeLeases[address] / amountTotalLeased);
		var payout = true;
	} else {
		var share = (activeLeases[address] / amountTotalLeased);
		var payout = false;
	  }

        var amount = fee * share;

       	if (address in payments) {
       		payments[address] += amount * (config.percentageOfFeesToDistribute / 100);
	} else {
		payments[address] = amount * (config.percentageOfFeesToDistribute / 100);
	}

	if ( payout == true ) {
        	console.log(address + ' will receive ' + amount + ' of(' + fee + ') share: ' + share);
	} else if ( payout == false ) {
		console.log(address + ' marked as NOPAYOUT, will not receive fee share.');
	}
    }
};

/**
 * Method that creates the concrete payment tx and writes it to the file
 * configured in the config section.
 */
var pay = function() {
    var transactions = [];
    var totalfees =0;

    var html = "";

    var html = "<!DOCTYPE html>" +
"<html lang=\"en\">" +
"<head>" +
"  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">" +
"  <link rel=\"stylesheet\" href=\"https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css\">" +
"  <script src=\"https://code.jquery.com/jquery-3.3.1.min.js\"></script>" +
"  <script src=\"https://stackpath.bootstrapcdn.com/bootstrap/4.2.1/js/bootstrap.min.js\"></script>" +
"</head>" +
"<body>" +

"<div class=\"container\">" +
"  <h3>Fee's between blocks " + config.startBlockHeight + " - " + config.endBlock + ", Payout #" + config.paymentid + "</h3>" +
"  <h4>(LPOS address: " + config.address + ")</h4>" +
"  <h5>Blocks forged: " + BlockCount + "</h5>" +
"  <table class=\"table table-striped table-hover pt-5\">" +
"    <thead> " +
"      <tr>" +
"        <th>Address</th>" +
"        <th>LTO</th>" +
"      </tr>" +
"    </thead>" +
"    <tbody>";

    for (var address in payments) {
        var payment = (payments[address] / Math.pow(10, 8));

	if ( nofeearray.indexOf(address) == -1 ) {

		console.log(address + ' will receive ' + parseFloat(payment).toFixed(8) )

		//send lto fee
		if (Number(Math.round(payments[address])) > 0) {
			transactions.push({
				"amount": Number(Math.round(payments[address])),
				"fee": config.feeAmount,
				"sender": config.address,
				"recipient": address
			});
		}

	} else {
		console.log(address + ' marked as NOPAYOUT, will not receive fee share.')
	  }

        totalfees += payments[address];


        html += "<tr><td>" + address + "</td><td>" + 							 	//address column
				((payments[address]/100000000).toFixed(8)) + "</td></tr>" + 	//lto fee's
				"\r\n";
    }	//End for loop

    html += "<tr><td><b>Total</b></td><td><b>" + ((totalfees/100000000).toFixed(8)) + "</b></td></tr>" +
			"\r\n";

    html += "</tbody>" +
"  </table>" +
"</div>" +

"</body>" +
"</html>";

    console.log("total lto fees: " + (totalfees/100000000) );
    var paymentfile = config.filename + config.paymentid + ".json";
    var htmlfile = config.filename + config.paymentid + ".html";

//if ( !BlockCount == 0 ) { transactions.push( { "forgedblocks:": BlockCount } ) }

    fs.writeFile(paymentfile, JSON.stringify(transactions), {}, function(err) {
	if (!err) {
		console.log('Planned payments written to ' + paymentfile + '!');
	} else {
		console.log(err);
	  }
    });

    fs.writeFile(htmlfile, html, {}, function(err) {
	if (!err) {
		console.log('HTML written to ' + config.filename + config.paymentid  + '.html!');
	} else {
		console.log(err);
	  }
    });
   
    // Create logfile with paymentinfo for reference and troubleshooting 
    fs.writeFile(config.filename + config.paymentid + ".log", "total lto fees: " + (totalfees/100000000) + "\n"
	+ "Total blocks forged: " + BlockCount + "\n"
	+ "Payment ID of batch session: " + config.paymentid + "\n"
	+ "Payment startblock: " + paymentstartblock + "\n"
	+ "Payment stopblock: " + paymentstopblock + "\n"
	+ "Following addresses are skipped for payment; \n"
	+ JSON.stringify(nofeearray) + "\n", function(err) {
   	if (!err) {
        	    console.log('Summarized payoutinfo is written to ' + config.filename + config.paymentid + ".log!");
		    console.log();
        } else {
            console.log(err);
	  }
    	});
    // End create logfile

    var latestblockinfo = {};
    latestblockinfo["leases"]=myLeases;
    latestblockinfo["canceledleases"]=myCanceledLeases;
    var blockleases = config.endBlock + "_" + config.address + ".json" ;

    fs.writeFile(blockleases, JSON.stringify(latestblockinfo), {}, function(err) {
        if (!err) {
            console.log('Leaseinfo written to ' + blockleases + '!');
        } else {
            console.log(err);
        }
    });

    
    var ActiveLeaseData = getActiveLeasesAtBlock(LastBlock);

    fs.writeFile("LastBlockLeasers.json", JSON.stringify(ActiveLeaseData), {}, function(err) {
        if (!err) {
            console.log('ActiveLeasers written to LastBlockLeasers.json!');
        } else {
            console.log(err);
        }
    });
    
   // Write the current payid of the batch to the payment queue file. This is used by the masspayment tool
   let paymentqueue = function (callback) {

         payarray = [ ];

         if ( fs.existsSync(payqueuefile) == false ) {  //There is no paymentqueue file!

                console.log("\nApparently there's no paymentqueue file yet. Adding paymentid '" + payid + "' of current batch to queuefile " + payqueuefile);
		console.log("You can now either start the next collector session, when finished it will automatically be added to the payment queue.");
		console.log("You can also verify the payment queue with the payment check tool (checkPayment.js). All pending payments are");
		console.log("automatically found and checked.");
		console.log("Or you can run the payment tool (masspayment.js), which automatically executes the transactions to all leasers for all");
		console.log("jobs in the payment queue. Whenever a job is finished, it is automatically removed from the payment queue file.\n");

                payarray = [ payid ];

         } else {       // there is a paymentqueue file!

                rawarray = fs.readFileSync(payqueuefile, function(err, data)  { //read it into array
                        if (err) { console.log("\nWARNING! Error reading paymentqueue file. terminating tool. Run batch " + payid + " again.\n");return; }
                });
                payarray = JSON.parse(rawarray); //read it into array

                //case 1. It's empty
                if ( payarray.length == 0 ) {
                        console.log("\nCurrently there are no payments pending in the queue.");
                        console.log("Adding paymentid '" + payid + "' to queuefile " + payqueuefile + ". This is the only payment in the queue now :-)\n");
			console.log("You can now either start the next collector session, when finished it will automatically be added to the payment queue.");
			console.log("You can also verify the payment queue with the payment check tool (checkPayment.js). All pending payments are");
			console.log("automatically found and checked.");
			console.log("Or you can run the payment tool (masspayment.js), which automatically executes the transactions to all leasers for all");
			console.log("jobs in the payment queue. Whenever a job is finished, it is automatically removed from the payment queue file.\n");
 
                        payarray = [ payid ]
                }
                //case 2. It's not empty, but has paymentid duplicates waiting
                else if ( payarray.includes (payid) == true ) {

                        console.log("\nWARNING! Found paymentid " + payid + " already in queue. This means there has already ran a batch with this id,\n"
                                   +"for which payments were not done yet. If you expect this because you used the old batchinfo file again, then it's fine.\n"
                                   +"The paymentqueue stays the same and has following payments waiting [" + payarray + "]. The batchinfo that was used in current run is:\n");
                        console.log(mybatchdata);
	                console.log("\nYou can verify the actual payments that will be done in a dry run first by starting the checkPaymentsFile.js script.");
                        console.log("\nHowever, if you weren't expecting a job with same paymentid in the queue (which normally shouldn't), then check logs!!!\n");
                }
                //case 3. It's not empty. Add current batch to queue
                else {
                        console.log("\nFound " + payarray.length + " pending payments already in queue. Adding current batch with paymentid " + payid + " to the queue.")
                        payarray.push(payid);
                        console.log("The total queue waiting for payouts is now: " + payarray);
                }

           }

	nextpayid = payid + 1
	console.log("The next batch session will be '" + nextpayid + "'\n");

	fs.writeFileSync(payqueuefile, JSON.stringify(payarray), function (err)  {
		if (err) {
			console.log("\nWARNING! Error updating payment queue file. Terminating tool. Run batch " + payid + " again.\n");
			return;
		}
   	});
   	callback();
   };

   // update json batchdata for next collection round
   let nextbatchdata = function () {

	mybatchdata["paymentid"] = (payid + 1).toString()
	mybatchdata["paystartblock"] = (paymentstopblock).toString()
	mybatchdata["paystopblock"] = (paymentstopblock + blockwindowsize).toString()
	
	fs.writeFile(batchinfofile, JSON.stringify(batchinfo), (err) => {
		if (err) {
			console.log("Something went wrong updating the file:" + batchinfofile + "!");
			console.log(err);
		} else {
			console.log("Batchinfo for next payment round is updated in file " + batchinfofile + "!");
			console.log();
	  	  }
    	});
    };

    // update the paymentqueue and callback update batchdata function
    paymentqueue(nextbatchdata);
};

/**
 * This method returns (block-exact) the active leases and the total amount
 * of leased lto for a given block.
 *
 * @param block the block to consider
 * @returns {{totalLeased: number, activeLeases: {}}} total amount of leased lto and active leases for the given block
 */
var getActiveLeasesAtBlock = function(block) {
    var activeLeases = [];
    var totalLeased = 0;
    var activeLeasesPerAddress = {};

    for (var leaseId in myLeases) {
        var currentLease = myLeases[leaseId];

        if (!myCanceledLeases[leaseId] || myCanceledLeases[leaseId].block > block.height) {
            activeLeases.push(currentLease);
        }
    }
    activeLeases.forEach(function (lease) {
        if (block.height > lease.block + 1000) {
            if (!activeLeasesPerAddress[lease.sender]) {
                activeLeasesPerAddress[lease.sender] = lease.amount;
            } else {
                activeLeasesPerAddress[lease.sender] += lease.amount;
            }

            totalLeased += lease.amount;
        }
    });

    return { totalLeased: totalLeased, activeLeases: activeLeasesPerAddress };
};

start();
